File: v4/rust/iprange-livedb/src/draft_store/timestamp_refresh.rs
//! First-seen and last-seen semantics over the authoritative ordered range merge.

use std::mem::MaybeUninit;

use crate::bootstrap::Bootstrap;
use crate::cancellation::CancellationToken;
use crate::cardinality::Cardinality129;
use crate::contract::AddressFamily;
use crate::error::{Error, Result};
use crate::key::{IpKey, Ipv4Key, Ipv6Key};
use crate::workflow::{Comparison, FirstSeenRemoval, FirstSeenRemovalSink};

use super::range_merge::{self, MapComparison, Policy};
use super::DraftStore;

pub(crate) struct TimestampMerge {
    pub(crate) input_intervals: u64,
    pub(crate) input_addresses: Cardinality129,
    pub(crate) comparison: Comparison,
}

impl DraftStore<'_> {
    pub(crate) fn merge_first_seen(
        &mut self,
        base: &Bootstrap,
        refresh_value: u32,
        cancellation: &CancellationToken,
    ) -> Result<TimestampMerge> {
        self.merge_timestamp(
            base,
            FirstSeenPolicy::new(refresh_value, NoRemovals),
            cancellation,
        )
    }

    pub(crate) fn merge_first_seen_with_removals<K, S>(
        &mut self,
        base: &Bootstrap,
        refresh_value: u32,
        sink: &mut S,
        cancellation: &CancellationToken,
    ) -> Result<TimestampMerge>
    where
        K: IpKey,
        S: FirstSeenRemovalSink<K>,
    {
        if base.meta.address_family != K::FAMILY {
            return Err(Error::WrongAddressFamily(
                "removal sink family does not match the database",
            ));
        }
        self.merge_timestamp_family::<K, _>(
            base,
            FirstSeenPolicy::new(refresh_value, BatchedRemovals::new(sink)),
            cancellation,
        )
    }

    pub(crate) fn merge_last_seen(
        &mut self,
        base: &Bootstrap,
        refresh_value: u32,
        cutoff: u32,
        cancellation: &CancellationToken,
    ) -> Result<TimestampMerge> {
        self.merge_timestamp(
            base,
            LastSeenPolicy::new(refresh_value, cutoff),
            cancellation,
        )
    }

    fn merge_timestamp<P>(
        &mut self,
        base: &Bootstrap,
        policy: P,
        cancellation: &CancellationToken,
    ) -> Result<TimestampMerge>
    where
        P: Policy<Ipv4Key, (), Output = TimestampOutput>
            + Policy<Ipv6Key, (), Output = TimestampOutput>,
    {
        match base.meta.address_family {
            AddressFamily::Ipv4 => {
                self.merge_timestamp_family::<Ipv4Key, _>(base, policy, cancellation)
            }
            AddressFamily::Ipv6 => {
                self.merge_timestamp_family::<Ipv6Key, _>(base, policy, cancellation)
            }
        }
    }

    fn merge_timestamp_family<K, P>(
        &mut self,
        base: &Bootstrap,
        policy: P,
        cancellation: &CancellationToken,
    ) -> Result<TimestampMerge>
    where
        K: IpKey,
        P: Policy<K, (), Output = TimestampOutput>,
    {
        let input_meta = self.draft.meta;
        let (input_intervals, finished) = range_merge::merge_coverage::<K, _>(
            self,
            &input_meta,
            base,
            policy,
            cancellation,
            "timestamp input intervals",
        )?;
        Ok(TimestampMerge {
            input_intervals,
            input_addresses: finished.output.input_addresses,
            comparison: finished.output.comparison,
        })
    }
}

struct TimestampOutput {
    input_addresses: Cardinality129,
    comparison: Comparison,
}

#[derive(Default)]
struct TimestampCounters {
    input_addresses: Cardinality129,
    comparison: MapComparison,
}

impl TimestampCounters {
    fn observe<K: IpKey>(
        &mut self,
        from: K,
        to: K,
        old: Option<u32>,
        incoming: Option<()>,
        new: Option<u32>,
    ) -> Result<Cardinality129> {
        let count = from.inclusive_cardinality(to)?;
        if incoming.is_some() {
            self.input_addresses = range_merge::add(self.input_addresses, count)?;
        }
        self.comparison.observe_count(count, old, new)?;
        Ok(count)
    }

    fn finish(self) -> Result<TimestampOutput> {
        Ok(TimestampOutput {
            input_addresses: self.input_addresses,
            comparison: self.comparison.finish()?,
        })
    }
}

struct FirstSeenPolicy<O> {
    refresh_value: u32,
    counters: TimestampCounters,
    removals: O,
}

impl<O> FirstSeenPolicy<O> {
    fn new(refresh_value: u32, removals: O) -> Self {
        Self {
            refresh_value,
            counters: TimestampCounters::default(),
            removals,
        }
    }
}

impl<K, O> Policy<K, ()> for FirstSeenPolicy<O>
where
    K: IpKey,
    O: RemovalObserver<K>,
{
    type Output = TimestampOutput;

    fn transform(
        &mut self,
        _store: &mut DraftStore<'_>,
        old: Option<u32>,
        incoming: Option<()>,
    ) -> Result<Option<u32>> {
        Ok(match (old, incoming) {
            (Some(value), Some(())) => Some(value),
            (None, Some(())) => Some(self.refresh_value),
            (_, None) => None,
        })
    }

    fn observe(
        &mut self,
        from: K,
        to: K,
        old: Option<u32>,
        incoming: Option<()>,
        new: Option<u32>,
    ) -> Result<()> {
        let addresses = self.counters.observe(from, to, old, incoming, new)?;
        if let (Some(first_seen), None) = (old, incoming) {
            self.removals.push(FirstSeenRemoval {
                from,
                to,
                first_seen,
                addresses,
            })?;
        }
        Ok(())
    }

    fn finish(self) -> Result<Self::Output> {
        self.removals.finish()?;
        self.counters.finish()
    }
}

trait RemovalObserver<K> {
    fn push(&mut self, removal: FirstSeenRemoval<K>) -> Result<()>;
    fn finish(self) -> Result<()>;
}

struct NoRemovals;

impl<K> RemovalObserver<K> for NoRemovals {
    #[inline]
    fn push(&mut self, _removal: FirstSeenRemoval<K>) -> Result<()> {
        Ok(())
    }

    #[inline]
    fn finish(self) -> Result<()> {
        Ok(())
    }
}

const REMOVAL_BATCH_CAPACITY: usize = 64;

struct BatchedRemovals<'a, K: Copy, S> {
    sink: &'a mut S,
    records: [MaybeUninit<FirstSeenRemoval<K>>; REMOVAL_BATCH_CAPACITY],
    len: usize,
}

impl<'a, K: Copy, S> BatchedRemovals<'a, K, S> {
    fn new(sink: &'a mut S) -> Self {
        Self {
            sink,
            records: [MaybeUninit::uninit(); REMOVAL_BATCH_CAPACITY],
            len: 0,
        }
    }

    fn flush(&mut self) -> Result<()>
    where
        S: FirstSeenRemovalSink<K>,
    {
        if self.len == 0 {
            return Ok(());
        }
        // SAFETY: the prefix is initialized by `push`, all records are Copy,
        // and the borrowed slice cannot outlive this synchronous call.
        let batch = unsafe {
            std::slice::from_raw_parts(
                self.records.as_ptr().cast::<FirstSeenRemoval<K>>(),
                self.len,
            )
        };
        self.sink.removals(batch)?;
        self.len = 0;
        Ok(())
    }
}

impl<K, S> RemovalObserver<K> for BatchedRemovals<'_, K, S>
where
    K: Copy,
    S: FirstSeenRemovalSink<K>,
{
    fn push(&mut self, removal: FirstSeenRemoval<K>) -> Result<()> {
        self.records[self.len].write(removal);
        self.len += 1;
        if self.len == REMOVAL_BATCH_CAPACITY {
            self.flush()?;
        }
        Ok(())
    }

    fn finish(mut self) -> Result<()> {
        self.flush()
    }
}

struct LastSeenPolicy {
    refresh_value: u32,
    cutoff: u32,
    counters: TimestampCounters,
}

impl LastSeenPolicy {
    fn new(refresh_value: u32, cutoff: u32) -> Self {
        Self {
            refresh_value,
            cutoff,
            counters: TimestampCounters::default(),
        }
    }
}

impl<K: IpKey> Policy<K, ()> for LastSeenPolicy {
    type Output = TimestampOutput;

    fn transform(
        &mut self,
        _store: &mut DraftStore<'_>,
        old: Option<u32>,
        incoming: Option<()>,
    ) -> Result<Option<u32>> {
        Ok(match (old, incoming) {
            (Some(value), Some(())) => Some(value.max(self.refresh_value)),
            (None, Some(())) => Some(self.refresh_value),
            (Some(value), None) if value > self.cutoff => Some(value),
            _ => None,
        })
    }

    fn observe(
        &mut self,
        from: K,
        to: K,
        old: Option<u32>,
        incoming: Option<()>,
        new: Option<u32>,
    ) -> Result<()> {
        self.counters
            .observe(from, to, old, incoming, new)
            .map(|_| ())
    }

    fn finish(self) -> Result<Self::Output> {
        self.counters.finish()
    }
}

File: v4/go/internal/writer/timestamp_refresh.go
// First-seen and last-seen refresh semantics over the authoritative
// ordered range merge (Rust draft_store/timestamp_refresh.rs parity):
// the workflow coverage tree built by the constant-range input is merged
// over the committed base through the timestamp policies, producing the
// exact input-address count and the before/after value classification.

package writer

import (
	"github.com/firehol/iprange/v4/go/internal/format"
)

// TimestampMerge is the outcome of one first-seen or last-seen refresh
// merge (Rust TimestampMerge).
type TimestampMerge struct {
	InputIntervals uint64
	InputAddresses format.Cardinality129
	Comparison     Comparison
}

// timestampOutput is the finished policy output of one timestamp merge
// (Rust TimestampOutput).
type timestampOutput struct {
	inputAddresses format.Cardinality129
	comparison     Comparison
}

// timestampCounters accumulates the input-address count and the
// before/after value classification of one timestamp merge (Rust
// TimestampCounters over MapComparison). The comparison balances at
// finish exactly like the Rust MapComparison::finish.
type timestampCounters[K any] struct {
	inputAddresses format.Cardinality129
	comparison     Comparison
	codec          rangeFamily[K]
}

func (c *timestampCounters[K]) observe(from, to K, old, incoming, new optionalValue) (format.Cardinality129, error) {
	count, err := familyInclusiveCardinalityOf(c.codec, from, to)
	if err != nil {
		return format.CardinalityZero(), err
	}
	if incoming.present {
		c.inputAddresses, err = c.inputAddresses.Add(count)
		if err != nil {
			return format.CardinalityZero(), overflow("ordered merge address count")
		}
	}
	if old.present {
		c.comparison.Before, err = c.comparison.Before.Add(count)
		if err != nil {
			return format.CardinalityZero(), overflow("ordered merge address count")
		}
	}
	if new.present {
		c.comparison.After, err = c.comparison.After.Add(count)
		if err != nil {
			return format.CardinalityZero(), overflow("ordered merge address count")
		}
	}
	switch {
	case old.present && new.present && old.value == new.value:
		c.comparison.Unchanged, err = c.comparison.Unchanged.Add(count)
	case old.present && new.present:
		c.comparison.Changed, err = c.comparison.Changed.Add(count)
	case old.present:
		c.comparison.Removed, err = c.comparison.Removed.Add(count)
	case new.present:
		c.comparison.Added, err = c.comparison.Added.Add(count)
	}
	if err != nil {
		return format.CardinalityZero(), overflow("ordered merge address count")
	}
	return count, nil
}

// finish balances the classification like Rust MapComparison::finish:
// the before and after totals recomputed from the classes must equal the
// directly counted totals.
func (c *timestampCounters[K]) finish() (timestampOutput, error) {
	before, err := c.comparison.Unchanged.Add(c.comparison.Changed)
	if err != nil {
		return timestampOutput{}, overflow("ordered merge address count")
	}
	before, err = before.Add(c.comparison.Removed)
	if err != nil {
		return timestampOutput{}, overflow("ordered merge address count")
	}
	after, err := c.comparison.Unchanged.Add(c.comparison.Changed)
	if err != nil {
		return timestampOutput{}, overflow("ordered merge address count")
	}
	after, err = after.Add(c.comparison.Added)
	if err != nil {
		return timestampOutput{}, overflow("ordered merge address count")
	}
	if before.Compare(c.comparison.Before) != 0 || after.Compare(c.comparison.After) != 0 {
		return timestampOutput{}, corrupt("ordered merge counters do not balance")
	}
	return timestampOutput{inputAddresses: c.inputAddresses, comparison: c.comparison}, nil
}

// firstSeenPolicy preserves old values on covered addresses and stamps
// the refresh value on newly covered ones; uncovered addresses are
// removed and streamed through the removal observer (Rust
// FirstSeenPolicy over RemovalObserver).
type firstSeenPolicy[K any] struct {
	refreshValue uint32
	counters     timestampCounters[K]
	removals     removalObserver[K]
}

func newFirstSeenPolicy[K any](refreshValue uint32, codec rangeFamily[K]) *firstSeenPolicy[K] {
	return &firstSeenPolicy[K]{refreshValue: refreshValue, counters: timestampCounters[K]{codec: codec}, removals: noRemovals[K]{}}
}

func newFirstSeenPolicyWithRemovals[K any](refreshValue uint32, codec rangeFamily[K], removals removalObserver[K]) *firstSeenPolicy[K] {
	return &firstSeenPolicy[K]{refreshValue: refreshValue, counters: timestampCounters[K]{codec: codec}, removals: removals}
}

func (p *firstSeenPolicy[K]) transform(store *DraftStore, old optionalValue, incoming incomingValue[uint32]) (optionalValue, error) {
	switch {
	case old.present && incoming.present:
		return old, nil
	case !old.present && incoming.present:
		return someValue(p.refreshValue), nil
	default:
		return noneValue(), nil
	}
}

func (p *firstSeenPolicy[K]) observe(from, to K, old optionalValue, incoming incomingValue[uint32], new optionalValue) error {
	addresses, err := p.counters.observe(from, to, old, optionalValue{value: incoming.value, present: incoming.present}, new)
	if err != nil {
		return err
	}
	if old.present && !incoming.present {
		return p.removals.push(firstSeenRemoval[K]{from: from, to: to, firstSeen: old.value, addresses: addresses})
	}
	return nil
}

func (p *firstSeenPolicy[K]) finish() (timestampOutput, error) {
	if err := p.removals.finish(); err != nil {
		return timestampOutput{}, err
	}
	return p.counters.finish()
}

func (p *firstSeenPolicy[K]) preserveWithoutInput() bool { return false }

// firstSeenRemoval is one first-seen interval removed by a refresh
// (Rust FirstSeenRemoval<K>).
type firstSeenRemoval[K any] struct {
	from, to  K
	firstSeen uint32
	addresses format.Cardinality129
}

// removalObserver consumes first-seen removals as the merge produces
// them (Rust RemovalObserver): push receives each removal, finish
// flushes the tail.
type removalObserver[K any] interface {
	push(removal firstSeenRemoval[K]) error
	finish() error
}

// noRemovals discards removals (Rust NoRemovals).
type noRemovals[K any] struct{}

func (noRemovals[K]) push(firstSeenRemoval[K]) error { return nil }
func (noRemovals[K]) finish() error                  { return nil }

// firstSeenRemovalBatch is the bounded removal batch size (Rust
// REMOVAL_BATCH_CAPACITY).
const firstSeenRemovalBatch = 64

// batchedRemovals4 buffers first-seen removals into bounded IPv4
// batches and flushes them through the sink (Rust BatchedRemovals over
// FirstSeenRemoval<Ipv4Key>). The batch slice is borrowed for the
// synchronous sink call; the sink must not retain it.
type batchedRemovals4 struct {
	sink    FirstSeenRemoval4Sink
	records [firstSeenRemovalBatch]FirstSeenRemoval4
	length  int
}

func (b *batchedRemovals4) push(removal firstSeenRemoval[key4]) error {
	b.records[b.length] = FirstSeenRemoval4{From: uint32(removal.from), To: uint32(removal.to), FirstSeen: removal.firstSeen, Addresses: removal.addresses}
	b.length++
	if b.length == firstSeenRemovalBatch {
		return b.flush()
	}
	return nil
}

func (b *batchedRemovals4) flush() error {
	if b.length == 0 {
		return nil
	}
	if err := b.sink(b.records[:b.length]); err != nil {
		return err
	}
	b.length = 0
	return nil
}

func (b *batchedRemovals4) finish() error { return b.flush() }

// batchedRemovals6 buffers first-seen removals into bounded IPv6
// batches (Rust BatchedRemovals over FirstSeenRemoval<Ipv6Key>).
type batchedRemovals6 struct {
	sink    FirstSeenRemoval6Sink
	records [firstSeenRemovalBatch]FirstSeenRemoval6
	length  int
}

func (b *batchedRemovals6) push(removal firstSeenRemoval[key6]) error {
	b.records[b.length] = FirstSeenRemoval6{FromHi: removal.from.Hi, FromLo: removal.from.Lo, ToHi: removal.to.Hi, ToLo: removal.to.Lo, FirstSeen: removal.firstSeen, Addresses: removal.addresses}
	b.length++
	if b.length == firstSeenRemovalBatch {
		return b.flush()
	}
	return nil
}

func (b *batchedRemovals6) flush() error {
	if b.length == 0 {
		return nil
	}
	if err := b.sink(b.records[:b.length]); err != nil {
		return err
	}
	b.length = 0
	return nil
}

func (b *batchedRemovals6) finish() error { return b.flush() }

// lastSeenPolicy refreshes covered values to at least the refresh value,
// keeps recent absence, and expires absence at or below the cutoff (Rust
// LastSeenPolicy).
type lastSeenPolicy[K any] struct {
	refreshValue uint32
	cutoff       uint32
	counters     timestampCounters[K]
}

func newLastSeenPolicy[K any](refreshValue, cutoff uint32, codec rangeFamily[K]) *lastSeenPolicy[K] {
	return &lastSeenPolicy[K]{
		refreshValue: refreshValue,
		cutoff:       cutoff,
		counters:     timestampCounters[K]{codec: codec},
	}
}

func (p *lastSeenPolicy[K]) transform(store *DraftStore, old optionalValue, incoming incomingValue[uint32]) (optionalValue, error) {
	switch {
	case old.present && incoming.present:
		if old.value >= p.refreshValue {
			return old, nil
		}
		return someValue(p.refreshValue), nil
	case !old.present && incoming.present:
		return someValue(p.refreshValue), nil
	case old.present && old.value > p.cutoff:
		return old, nil
	default:
		return noneValue(), nil
	}
}

func (p *lastSeenPolicy[K]) observe(from, to K, old optionalValue, incoming incomingValue[uint32], new optionalValue) error {
	_, err := p.counters.observe(from, to, old, optionalValue{value: incoming.value, present: incoming.present}, new)
	return err
}

func (p *lastSeenPolicy[K]) finish() (timestampOutput, error) {
	return p.counters.finish()
}

func (p *lastSeenPolicy[K]) preserveWithoutInput() bool { return false }

// mergeFirstSeen merges the workflow coverage tree over the committed
// base with the first-seen policy (Rust DraftStore::merge_first_seen).
func (s *DraftStore) mergeFirstSeen(base format.Meta, refreshValue uint32, check func() error) (TimestampMerge, error) {
	if base.AddressFamily == format.AddressFamilyIPv4 {
		return mergeTimestamp(s, rangeCodec4{}, base, newFirstSeenPolicy(refreshValue, rangeCodec4{}), check)
	}
	return mergeTimestamp(s, rangeCodec6{}, base, newFirstSeenPolicy(refreshValue, rangeCodec6{}), check)
}

// mergeFirstSeenWithRemovals4 merges with the first-seen policy and
// streams every removed interval through the IPv4 sink in bounded
// batches (Rust DraftStore::merge_first_seen_with_removals over
// Ipv4Key). The sink family must match the database family (Rust
// K::FAMILY check); a mismatch reports the exact Rust error class and
// detail.
func (s *DraftStore) mergeFirstSeenWithRemovals4(base format.Meta, refreshValue uint32, sink FirstSeenRemoval4Sink, check func() error) (TimestampMerge, error) {
	if base.AddressFamily != format.AddressFamilyIPv4 {
		return TimestampMerge{}, &format.Error{Code: format.CodeWrongAddressFamily, Detail: "removal sink family does not match the database"}
	}
	return mergeTimestamp(s, rangeCodec4{}, base, newFirstSeenPolicyWithRemovals(refreshValue, rangeCodec4{}, &batchedRemovals4{sink: sink}), check)
}

// mergeFirstSeenWithRemovals6 merges with the first-seen policy and
// streams every removed interval through the IPv6 sink in bounded
// batches (Rust DraftStore::merge_first_seen_with_removals over
// Ipv6Key).
func (s *DraftStore) mergeFirstSeenWithRemovals6(base format.Meta, refreshValue uint32, sink FirstSeenRemoval6Sink, check func() error) (TimestampMerge, error) {
	if base.AddressFamily != format.AddressFamilyIPv6 {
		return TimestampMerge{}, &format.Error{Code: format.CodeWrongAddressFamily, Detail: "removal sink family does not match the database"}
	}
	return mergeTimestamp(s, rangeCodec6{}, base, newFirstSeenPolicyWithRemovals(refreshValue, rangeCodec6{}, &batchedRemovals6{sink: sink}), check)
}

// mergeLastSeen merges the workflow coverage tree over the committed
// base with the last-seen policy (Rust DraftStore::merge_last_seen).
func (s *DraftStore) mergeLastSeen(base format.Meta, refreshValue, cutoff uint32, check func() error) (TimestampMerge, error) {
	if base.AddressFamily == format.AddressFamilyIPv4 {
		return mergeTimestamp(s, rangeCodec4{}, base, newLastSeenPolicy(refreshValue, cutoff, rangeCodec4{}), check)
	}
	return mergeTimestamp(s, rangeCodec6{}, base, newLastSeenPolicy(refreshValue, cutoff, rangeCodec6{}), check)
}

// mergeTimestamp drives the timestamp policy through the coverage merge
// over the draft input tree and the committed base (Rust
// merge_timestamp_family): the draft meta is the input, the merge output
// replaces the draft range root, and the base tree retires inside the
// merge finish.
func mergeTimestamp[K any](s *DraftStore, codec rangeFamily[K], base format.Meta, policy mergePolicy[uint32, timestampOutput, K], check func() error) (TimestampMerge, error) {
	inputIntervals, finished, err := mergeCoverage[timestampOutput, K](s, s.draft.meta, base, codec, policy, check, "timestamp input intervals")
	if err != nil {
		return TimestampMerge{}, err
	}
	return TimestampMerge{
		InputIntervals: inputIntervals,
		InputAddresses: finished.inputAddresses,
		Comparison:     finished.comparison,
	}, nil
}

// FirstSeenRemoval4 is one IPv4 first-seen interval removed by a
// complete refresh (Rust FirstSeenRemoval<Ipv4Key>).
type FirstSeenRemoval4 struct {
	From, To  uint32
	FirstSeen uint32
	Addresses format.Cardinality129
}

// FirstSeenRemoval6 is one IPv6 first-seen interval removed by a
// complete refresh (Rust FirstSeenRemoval<Ipv6Key>).
type FirstSeenRemoval6 struct {
	FromHi, FromLo, ToHi, ToLo uint64
	FirstSeen                  uint32
	Addresses                  format.Cardinality129
}

// FirstSeenRemoval4Sink receives bounded batches of IPv4 first-seen
// removals (Rust FirstSeenRemovalSink<Ipv4Key>). The batch slice is
// borrowed for the synchronous call; the sink must not retain it.
type FirstSeenRemoval4Sink func(removals []FirstSeenRemoval4) error

// FirstSeenRemoval6Sink receives bounded batches of IPv6 first-seen
// removals (Rust FirstSeenRemovalSink<Ipv6Key>). The batch slice is
// borrowed for the synchronous call; the sink must not retain it.
type FirstSeenRemoval6Sink func(removals []FirstSeenRemoval6) error
